// component-library/src/Collection/NamedNodeArray.ts
export default class NamedNodeArray extends Array<Attr> implements NamedNodeMap {
    public getNamedItem(qualifiedName: string): Attr {
        return this.find(node => node.name === qualifiedName);
    }
    public getNamedItemNS(namespace: string, localName: string): Attr {
        return this.find(node => node.namespaceURI === namespace && node.localName === localName);
    }
    public item(index: number): Attr {
        return this[index];
    }
    public removeNamedItem(qualifiedName: string): Attr {
        const node = this.getNamedItem(qualifiedName);
        this.remove(node);
        return node;
    }
    public removeNamedItemNS(namespace: string, localName: string): Attr {
        const node = this.getNamedItemNS(namespace, localName);
        this.remove(node);
        return node;
    }
    public setNamedItem(attr: Attr): Attr {
        this.push(attr);
        return attr;
    }
    public setNamedItemNS(attr: Attr): Attr {
        this.push(attr);
        return attr;
    }
}
